cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -2355,6 +2358,98 @@ const FunctionDoc array_take_doc(
      "given by `indices`.  Nulls in `indices` emit null in the output."),
     {"array", "indices"}, "TakeOptions");
 
+const FunctionDoc indices_nonzero_doc(
+    "Return the indices of the values in the array that are non-zero",
+    ("For each input value, check if it's zero, false or null. Emit the index\n"
+     "of the value in the array if it's none of the those."),
+    {"values"});
+
+struct NonZeroVisitor {
+  UInt64Builder* builder;
+  const ArrayDataVector& arrays;
+
+  NonZeroVisitor(UInt64Builder* builder, const ArrayDataVector& arrays)
+      : builder(builder), arrays(arrays) {}
+
+  Status Visit(const DataType& type) { return Status::NotImplemented(type.ToString()); }
+
+  template <typename Type>
+  enable_if_t<is_primitive_ctype<Type>::value, Status> Visit(const Type&) {
+    using T = typename GetViewType<Type>::T;
+    uint32_t index = 0;
+
+    for (ArrayDataVector::const_iterator current_array = arrays.begin();
+         current_array != arrays.end(); ++current_array) {
```

Comment:
Modern C++ allows you to spell this out much more conveniently:
```c++
for (const auto& current_array : arrays) {
   ...
```

:-)

Reply:
right! Sorry, not yet used to range based iteration, my muscle memory still goes to interators as the first thing :D